src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperAttributes.cs

```diff
@@ -0,0 +1,149 @@
+
+        public TagHelperAttribute<TAttributeValue> this[int index]
+        {
+            get
+            {
+                return _attributes[index];
+            }
+            set
+            {
+                _attributes[index] = value;
+            }
+        }
+
+        public bool TryGetAttribute([NotNull] string key, out TagHelperAttribute<TAttributeValue> attribute)
+        {
+            attribute = this[key];
+
+            return attribute != null;
+        }
+
+        public bool TryGetAttributes([NotNull] string key, out IEnumerable<TagHelperAttribute<TAttributeValue>> attributes)
+        {
+            attributes = _attributes.Where(attribute => MatchesAttribute(key, attribute));
+
+            return attributes.Any();
+        }
+
+        public void Add([NotNull] string key)
+        {
+            _attributes.Add(new TagHelperAttribute<TAttributeValue>(key)
+            {
+                Minimized = true
+            });
+        }
+
+        public void Add([NotNull] string key, TAttributeValue value)
+        {
+            _attributes.Add(new TagHelperAttribute<TAttributeValue>(key, value));
+        }
+
+        public void Add([NotNull] TagHelperAttribute<TAttributeValue> attribute)
+        {
+            _attributes.Add(attribute);
+        }
+
+        public void Insert(int index, TagHelperAttribute<TAttributeValue> item)
+        {
+            _attributes.Insert(index, item);
+        }
+
+        public void CopyTo([NotNull] TagHelperAttribute<TAttributeValue>[] attributes, int index)
+        {
+            _attributes.CopyTo(attributes, index);
+        }
+
+        public bool Remove([NotNull] string key)
+        {
+            IEnumerable<TagHelperAttribute<TAttributeValue>> attributes;
+            if (TryGetAttributes(key, out attributes))
+            {
+                foreach (var attribute in attributes)
+                {
+                    Remove(attribute);
+                }
+
+                return true;
+            }
+
+            return false;
+        }
+
+        public bool Remove([NotNull] TagHelperAttribute<TAttributeValue> item)
+        {
+            // REVIEW: Do we want this to be a comparer remove? If so it'd remove more than one.
+            return _attributes.Remove(item);
+        }
+
+        public void RemoveAt(int index)
+        {
+            _attributes.RemoveAt(index);
+        }
+
+        public void Clear()
+        {
+            _attributes.Clear();
+        }
+
+        public bool Contains([NotNull] string key)
+        {
+            return _attributes.Any(attribute => MatchesAttribute(key, attribute));
+        }
+
+        public bool Contains([NotNull] TagHelperAttribute<TAttributeValue> item)
+        {
+            // REVIEW: Do we want this to be a comparer check?
+            return _attributes.Contains(item);
+        }
+
+        public int IndexOf([NotNull] TagHelperAttribute<TAttributeValue> item)
```

Comment:
this method would also pick up the correct comparer if `TagHelperAttribute<TAttributeValue>` implemented `IEquatable<T>`